### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by amirahaile**
https://rocky-brook-8140.herokuapp.com/

Team members: Amira, Brandi, Kyna, and Sally

This is for our completed bEtsy project. 

We are very much proud of how our flash errors appear under the appropriate fields and the conditional button displays (such as only being able to cancel an order before it has been shipped by the merchant).

We had some trouble writing specs. After multiple attempts and long hours we concluded that our time might be better spent completing other requirements. Also, heroku is still the enemy.

### CODE REVIEW

**Comment by jnf** [comment]
Hi! I am so sorry I did not review this PR over the weekend. :( I am doing so now.

File: app/controllers/order_items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    @order_item.destroy
+    @order_item.save
+    redirect_to order_path
+  end
+
+  def qty_decrease
+    @order_item.quantity -= 1
+    @order_item.save
+
+    redirect_to :back rescue redirect_to order_items_decrease_path
```
Comment (jnf) [inline app/controllers/order_items_controller.rb]: This redirect would be a GET request, but the `order_items_decrease` only works for POST requests. I'm not sure this is where you would want to redirect the user if `:back` doesn't exist.

File: app/controllers/order_items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def qty_increase
+    @order_item.quantity += 1
+    @order_item.save
+
+    redirect_to :back rescue redirect_to order_items_increase_path
```
Comment (jnf) [inline app/controllers/order_items_controller.rb]: Ditto my previous comment.

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    # check for requested amount of inventory & update before completing payment
+    update_inventory(@order)
+
+    @order.email = params[:order][:email]
+    @order.address1 = params[:order][:address1]
+    @order.address2 = params[:order][:address2]
+    @order.city = params[:order][:city]
+    @order.state = params[:order][:state]
+    @order.zipcode = params[:order][:zipcode]
+    @order.card_last_4 = params[:order][:card_number.split(//)][-1, 4].join(",")
```
Comment (jnf) [inline app/controllers/orders_controller.rb]: Two notes:
- this should probably be a model callback (`before_validation`, most likely) that ensures the data is in the right format.
- Strings can be indexed like arrays. I'm pretty sure `"1234567890"[-4,4] #=> "7890"`

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  private
+
+  def update_inventory(order)
+    order.order_items.each do |order_item|
+      if order_item.product.inventory >= order_item.quantity
+        # remove item from inventory
+        order_item.product.inventory -= order_item.quantity
+        order_item.product.update!
+      else
+        redirect_to :back rescue redirect_to order_path(@order), flash.now[:error] = "#{order_item.product} only has #{order_item.quantity} item in stock."
```
Comment (jnf) [inline app/controllers/orders_controller.rb]: I'm not sure the redirect :back is appropriate here; is there anywhere this should ever go that isn't `order_path`?

File: app/controllers/orders_controller.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  def show
+    @order_items = @order.order_items
+  end
+
+  def payment
+    # check for requested amount of inventory & update before completing payment
+    update_inventory(@order)
+
+    @order.email = params[:order][:email]
```
Comment (jnf) [inline app/controllers/orders_controller.rb]: Is there a reason you've used all this manual assignment versus doing mass assignment with `update`?

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  #   # redirect_to #order_cart_path
+  # end
+
+  def show
+    @order_items = @order.order_items
+  end
+
+  def payment
+    # check for requested amount of inventory & update before completing payment
+    update_inventory(@order)
```
Comment (jnf) [inline app/controllers/orders_controller.rb]: This method worries me. Looking at line 56 below, the `update!` call is going to mutate the database, but you don't know yet if the order is valid. There could be other reasons to not complete the order.

Consider refactoring/moving the logic in the `update_inventory` method to a custom validation (`item_in_stock` or similar) and a `before_save` callback (`decrement_inventory` or similar).

File: app/controllers/products_controller.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
+class ProductsController < ApplicationController
+  before_action :require_login, only: [:new, :create, :edit, :update, :destroy]
+
+  def self.model
```
Comment (jnf) [inline app/controllers/products_controller.rb]: What's the purpose of this method?

File: app/controllers/products_controller.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+class ProductsController < ApplicationController
+  before_action :require_login, only: [:new, :create, :edit, :update, :destroy]
```
Comment (jnf) [inline app/controllers/products_controller.rb]: Where is `require_login` defined?

File: app/models/order.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  # Associations
+  has_many :order_items
+
+  # Validations
+  validates_presence_of :email, :address1, :city, :state, :zipcode,
+                        :card_last_4, :card_exp, :status
+  validates_length_of :state, is: 2, message: "must be state abbreviation" # must be two (capital) characters? ex. WA
+  validates_length_of :card_last_4, is: 4
+  validates_length_of :zipcode, within: 5..10
+  validates_format_of :card_last_4, with: /[0-9]{4}/, message: "only numbers allowed"
```
Comment (jnf) [inline app/models/order.rb:14]: Nice regex!

File: app/models/order.rb [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  validates_inclusion_of :status, in: %w(pending paid cancelled complete),
+                                  message: "%{value} is not a valid status"
+
+  # Callbacks
```
Comment (jnf) [inline app/models/order.rb]: Why is all of this commented out?

File: app/models/order.rb [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  #     "WA" => "Washington",
+  #     "WV" => "West Virginia",
+  #     "WI" => "Wisconsin",
+  #     "WY" => "Wyoming"
+  #   }
+  #   return if self.state == nil
+  #   return if states.keys.includes?(self.state)
+  #
+  #   if states.values.includes?(self.state)
+  #     self.state = states.find { |abbr, full| full == self.state }[0]
```
Comment (jnf) [inline app/models/order.rb:97]: Presenting states as a select box might remove the necessity for this kind of scrubbing.

File: app/models/product.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  validates :name, uniqueness: true
+  validates :price, numericality: { greater_than: 0 }
+  validates :inventory, numericality: { only_integer: true, greater_than: 0 }
+  validates_associated :user
+
+  # WORKING ON FOR PRODUCTS#INDEX: (DOES NOT WORK)
+  # scope :by_category, (category) -> { where(Product.all.categories.includes?(category)) }
+  # scope :by_category, -> (category) { where(categories: category) }
```
Comment (jnf) [inline app/models/product.rb]: If you're trying to get all the products in a _single_ category, the second scope is _almost_ there:

``` ruby
scope :by_category, -> (category) { where(category: category) }
```

File: app/views/orders/confirmation.html.erb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+<table class="table">
+  <tr>
+    <th>Name</th>
+    <th>Price</th>
+    <th>Qty</th>
+    <th>Total</th>
+  </tr>
+  <% @order_items.each do |item| %>
+  <tr>
+    <% product = Product.find(item.product_id) %>
```
Comment (jnf) [inline app/views/orders/confirmation.html.erb]: OrderItem :belongs_to :product, yeah? Skip doing this lookup and just call `@order_item.product` instead.

File: app/views/orders/confirmation.html.erb [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+		</td>
+	</tr>
+	<tr>
+		<td colspan="2">
+		</td>
+		<td>
+			Tax:
+		</td>
+		<td>
+			<% tax = subtotal * 0.095 %>
```
Comment (jnf) [inline app/views/orders/confirmation.html.erb:64]: The 0.095 is what we call a "magic number." This means it's a number that totally satisfies a requirement, but has no grounding in application logic; it's just a number appropriate for a specific situation. What if you have a client in a place that doesn't have sales tax? What if there's a state and a local sales tax? What if the sales tax is 7%?

There are no tax collection project requirements though, so I'm not gonna harp on it too much.

File: app/views/products/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  <%= @product.description %>
+
+  <p>Merchant: <%= @product.user.username %></p>
+  <p>
+    Categories: <% @product.categories.each do |cat| %>
+    <%= cat.name %>
+    <% end %>
+  </p>
+
+  <%= link_to "Add to Cart", new_order_path, { method: :get, id: @product.id, class: "btn btn-default" } %>
```
Comment (jnf) [inline app/views/products/show.html.erb]: Random note; `link_to` generates an anchor (`<a>`) tag, and, according to the HTTP spec, anchors always send GET requests by default, so the `method: :get` here is likely superfluous.

File: config/routes.rb [hunk 16]
```diff
@@ -0,0 +1,70 @@
+Rails.application.routes.draw do
+  # The priority is based upon order of creation: first created -> highest priority.
+  # See how all your routes lay out with "rake routes".
+
+  # You can have the root of your site routed with "root"
+  # root 'welcome#index'
+  root 'products#index'
+  resources :users, :products, :orders, :order_items, :reviews
```
Comment (jnf) [inline config/routes.rb]: `resources` will generate default routes that you may not want, specifically for the `reviews` workflow. Consider being more explicit with your routes.

**Comment by jnf** [comment]
Hi! Awesome job so far. I've left some comments and questions, so please have a look and let me know if you have any questions or concerns.

When yinz have gone through it all, please merge this PR. Keep up the good work!

File: app/models/order.rb (hunk 10 above)
Comment (QuartzBrandi) [inline app/models/order.rb]: Because it ain't working yet. ^-^; It's a WIP!

File: app/controllers/products_controller.rb (hunk 7 above)
Comment (QuartzBrandi) [inline app/controllers/products_controller.rb]: We believe it's required for the shared controller specs to work properly. Is this correct? Is this a place where a comment would be recommended?

File: app/controllers/products_controller.rb (hunk 8 above)
Comment (QuartzBrandi) [inline app/controllers/products_controller.rb]: It's a WIP and a reminder that we will need to require the user to be logged in for certain product actions.

File: app/models/order.rb (hunk 11 above)
Comment (QuartzBrandi) [inline app/models/order.rb:97]: Would we still want to validate for this or no?

File: config/routes.rb (hunk 16 above)
Comment (QuartzBrandi) [inline config/routes.rb]: Yeah, we just did that at the beginning so we could start working and plan to prune after we get the controllers all mapped out.

File: app/views/orders/confirmation.html.erb (hunk 14 above)
Comment (amirahaile) [inline app/views/orders/confirmation.html.erb:64]: How would this be handled IRL? Curious 😊

**Pull request merged**